tx.go

```diff
@@ -240,6 +225,67 @@ func caller(n int) string {
 	return name
 }
 
+type commandWithError struct {
+	command interface{}
+	errCh   chan error
+}
+type reportErrorCommand struct{ exceptionType, errorMessage, stackTrace, stackFrameDelim string }
+type beginGenericSegmentCommand struct{ name string }
+type beginDatastoreSegmentCommand struct{ table, operation, sql, rollupName string }
+type beginExternalSegmentCommand struct{ host, name string }
+type endSegmentCommand struct{}
+type endTransactionCommand struct{}
+
+func (t *tx) processCommands() {
+	for cmdWithErr := range t.ch {
+		cmd, errCh := cmdWithErr.command, cmdWithErr.errCh
+
+		switch cmd := (cmd).(type) {
+		case reportErrorCommand:
+			_, err := t.Reporter.ReportError(t.id, cmd.exceptionType, cmd.errorMessage, cmd.stackTrace, cmd.stackFrameDelim)
+			errCh <- err
+		case beginGenericSegmentCommand:
+			id, err := t.Tracer.BeginGenericSegment(t.id, t.ss.Peek(), cmd.name)
+			if err == nil {
+				t.ss.Push(id)
+			}
+			errCh <- err
+		case beginDatastoreSegmentCommand:
+			id, err := t.Tracer.BeginDatastoreSegment(t.id, t.ss.Peek(), cmd.table, cmd.operation, cmd.sql, cmd.rollupName)
+			if err == nil {
+				t.ss.Push(id)
+			}
+			errCh <- err
+		case beginExternalSegmentCommand:
+			id, err := t.Tracer.BeginExternalSegment(t.id, t.ss.Peek(), cmd.host, cmd.name)
+			if err == nil {
+				t.ss.Push(id)
+			}
+			errCh <- err
+		case endSegmentCommand:
+			if id, ok := t.ss.Pop(); ok {
+				err := t.Tracer.EndSegment(t.id, id)
+				errCh <- err
+			}
+		case endTransactionCommand:
+			for t.ss.Peek() != rootSegment {
+				t.EndSegment() // discarding errors?
+			}
+			err := t.Tracer.EndTransaction(t.id)
+			errCh <- err
+			close(t.ch)
```

Comment:
I've read it is bad practice for channel receivers to close channels. I'm not sure why it would matter in this case though.